lib/recursive-methods.rb

```diff
@@ -1,49 +1,88 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)?
+# Space complexity: O(n)?
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return n if n == 1
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+
+  n = n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ?
+# Space complexity: O(n) ?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return s[-1] + reverse(s[0...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ?
+# Space complexity: O(n) ?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return s[-1] + reverse(s[0...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)?
+# Space complexity: O(n)?
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return n if n == 0
+  return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)?
+# Space complexity: O(n) ?
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length == 0
+  return false if s.length % 2 != 0
+
+  if s[0] + s[-1] == "()"
+    return nested(s[1...-1])
+  else
+    return false
+  end
+
+  
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ?
+# Space complexity: O(n)?
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.length == 0
+  return true if value == array[0]
+  return search(array[1..-1], value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)?
+# Space complexity: O(n) ?
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length <= 1
+  if s[0] == s[-1]
+    return is_palindrome(s[1...-1])
+  else
+    return false
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)?
+# Space complexity: O(n)?
 def digit_match(n, m)
-    raise NotImplementedError, "Method not implemented"
+  n_dig = n.digits
+  m_dig = m.digits
+
+  min = [n_dig.length, m_dig.length].min
+
+  return 0 if (n_dig.length == 1 || m_dig.length == 1) && (n_dig[0] != m_dig[0])
+  return 1 if (n_dig.length == 1 || m_dig.length == 1) && (n_dig[0] == m_dig[0])
+  
+  if n_dig[0] == m_dig[0]
+    n = n_dig[1..(min - 1)].join.reverse.to_i
+    m = m_dig[1..(min - 1)].join.reverse.to_i
```

Comment:
I'm not sure why you're doing all the reversing and joining here.

Reply:
This was definitely not the most efficient way to do it.  I had to do this because digit creates a reverse array so I had to reverse it back to compare the correct indexes.